### DISCUSSION TITLE
seq on non-iterables

### DISCUSSION THREAD

**Discussion by brandonstubbs** [clavascript/clavascript#109] [Ideas]
Opening this to discuss how we want to handle `seq` with non-iterables.

At the moment when running a function that uses `seq`, map, first, first, etc. with a non-iterable, we get the following error:
```
# ./node_cli.js -e '(println (first 1))' 
file://.../clavascript/core.js:253
  return typeof x === 'string' || x === null || x === undefined || Symbol.iterator in x;
                                                                                   ^

TypeError: Cannot use 'in' operator to search for 'Symbol(Symbol.iterator)' in 1
    at seqable_QMARK_ (file:///.../clavascript/core.js:253:84)
...
```

Do we want to throw a controlled error out of `seq` [as clojurescript does](https://github.com/clojure/clojurescript/blob/master/src/main/cljs/cljs/core.cljs#L1253)

Or do we want to make the fix and let it fall back to `Object.entries` in the `seq` function
```javascript
function seqable_QMARK_(x) {
  // String is iterable but doesn't allow `m in s`
  return typeof x === 'string' || x === null || x === undefined || typeof x[Symbol.iterator] === 'function';
}
```
```bash
# ./node_cli.js -e '(println (first 1))'         
undefined
```

**Comment by lilactown** [comment]
I think that for primitives like numbers, we should throw an error. The contract is ideally:

1. You implement the `Symbol.iterator` protocol
2. You're a POJO
3. You're `nil`

Otherwise, `seq` should throw

**Comment by lilactown** [comment]
We should have a much more descriptive message in this case 😄

**Comment by borkdude** [comment]
I propose we start a Github Discussion in case we want to discuss and extract actionable information from that into issues, so those are clear units of work that can be started and finished with a clear scope.
